[3.3] Add validation between Placement Groups (PG) and On Demand Capacity Reservations (ODCRs)

dreambeyondorange opened the pull request:
Add validation between Placement Groups (PG) and On Demand Capacity Reservations (ODCRs)

Provides validation functions to ensure the PG supplied for each compute resource is compatible with the ODCR or group of ODCRs also provided for that resource.

If the PG is provided, there must either be an open, targeted, or PG/ODCR that matches

IF the PG is omitted, there must be an open or targeted ODCR that matches

Signed-off-by: Ryan Anderson <ndry@amazon.com>


### Description of changes
* See above

### Tests
* Manual tests for different combinations of PG and ODCR
* Unit tests for some failure scenarios and successful ones, not comprehensive

### References
* https://quip-amazon.com/kaLSAnnbyR7I/Design-Add-Support-for-On-Demand-Capacity-Reservations-ODCRs-in-Cluster-Configuration-File

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

EddyMM commented on cli/src/pcluster/validators/ec2_validators.py:
```diff
@@ -189,27 +192,111 @@ def _validate(self, capacity_reservation_id: str, instance_type: str, subnet: st
+    capacity_reservation_ids = AWSApi.instance().resource_groups.get_capacity_reservation_ids_from_group_resources(
+        capacity_reservation_resource_group_arn
+    )
+    print(capacity_reservation_ids)
```
Let's use logger.info(...) if we want to record which capacity reservation IDs were found, otherwise we can simply avoid logging it in the first place.

dreambeyondorange replied:
ty, I missed removing this print

EddyMM commented on cli/src/pcluster/validators/ec2_validators.py:
```diff
@@ -189,27 +192,111 @@ def _validate(self, capacity_reservation_id: str, instance_type: str, subnet: st
+                    capacity_reservation=odcr, instance_type=instance_type, subnet=queue.networking.subnet_ids[0]
+                ):
+                    odcr_pg = (
+                        get_resource_name_from_resource_arn(odcr["PlacementGroupArn"])
```
To simplify the logic we can use:

```
odcr_pg = get_resource_name_from_resource_arn(odcr.get("PlacementGroupArn", None))

if odcr_pg:
   ...
```

dreambeyondorange replied:
removed for this PR

dreambeyondorange replied:
ah yes thank you!

EddyMM reviewed (commented):
Empty file committed: [tests/networking/test_on_demand_capacity_reservation.py](https://github.com/aws/aws-parallelcluster/pull/4365/files#diff-89804e94190062eded2282947720400f686a5c17c55076424331f83011b0b161)

hanwen-pcluste commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1143,6 +1144,8 @@ def __init__(
         self.config_version = ""
         self.original_config_version = ""
         self._official_ami = None
+        # # Cache subnet ids information together to reduce number of boto3 calls
+        # AWSApi.instance().ec2.describe_subnets(subnet_ids=self.compute_subnet_ids)
```
Why this is commented?

dreambeyondorange replied:
I wanted to keep this visible, since it didn't work and I didn't have time to dive deeper right now.  Basically, the objects we need don't exist yet, so not sure if this is a viable path.

hanwen-pcluste commented on cli/src/pcluster/validators/ec2_validators.py:
```diff
@@ -189,27 +192,111 @@ def _validate(self, capacity_reservation_id: str, instance_type: str, subnet: st
+                        break
+                if not found_qualified_capacity_reservation:
+                    self._add_failure(
+                        f"Capacity reservation resource group {capacity_reservation_resource_group_arn} must have at "
+                        f"least one capacity reservation for {instance_type} in the same availability zone as subnet"
+                        f" {subnet}.",
```
How about adding more information to the message? Something like "... You can suppress the validator if you wish to launch this instance type without a capacity reservation."

dreambeyondorange committed:
Add validation between Placement Groups (PG) and On Demand Capacity Reservations (ODCRs)

Provides validation functions to ensure the PG supplied for each compute resource
is compatible with the ODCR or group of ODCRs also provided for that resource.

If the PG is provided, there must either be an open, targeted, or PG/ODCR that matches

If the PG is omitted, there must be an open or targeted ODCR that matches

Signed-off-by: Ryan Anderson <ndry@amazon.com>

hanwen-pcluste reviewed (commented):
Great work.

I left a few comments.

hanwen-pcluste reviewed (approved):
LGTM

hanwen-pcluste committed:
Merge branch 'develop' into develop
